### PR TITLE
Fix pickling issues on `MultiprocessIterator`

### PR DESCRIPTION
**Pull request by emcastillo**
When using `MultiprocessParallelUpdater` the `MultiprocessIterator` objects
are pickled when starting the workers.

The `MultiprocessIterator` object has an internal `_Communicator` object which holds a
threading.lock and is pickled too. However, this kind of object is not serializable with pickle.

This patch makes the `MultiprocessParallelUpdater` class to implement the pickle interface and relies on chainer `serialize` interface to implement object pickling.

### CODE REVIEW

**Comment by niboshi** [comment]
Background: https://github.com/chainer/chainer/issues/7309

**Comment by emcastillo** [comment]
Added tests, we need to wait for #7511 to be merged in order for the test to complete.

**Comment by emcastillo** [comment]
> It looks to me the issue is about `MultiprocessIterator` and it's independent of `MultiprocessParallelUpdater`. Is it not possible to write unit test using only `MultiprocessIterator`? Also, the fix to `MultiprocessParallelUpdater` would be better separated from this PR.

The pickling is done when combining MultiprocessIterator with MultiprocessParallelUpdater.
I can write a simple test that pickles the MultiprocessIterator without doing any training if you think its better.

**Comment by emcastillo** [comment]
I have started to split this PR: #7552 tackles the `multiprocessing` spawn check.
This PR is going to be refactored with pickle interface and pickle test.
Finally I will add another PR with the test of `MultiprocessIterator` and `MultiprocessParallelUpdater`

**Comment by emcastillo** [comment]
@niboshi now this PR is only about MultiprocessIterator and there is a test that tries to pickle it.

File: tests/chainer_tests/iterators_tests/test_multiprocess_iterator.py [hunk 1]
```diff
@@ -371,6 +372,34 @@ class TestMultiprocessIteratorSlow(
     pass
 
 
+# TODO (emcastill) : pickle fails when using order_sampler as lambda function
```
Comment (niboshi) [inline tests/chainer_tests/iterators_tests/test_multiprocess_iterator.py]: How about using a module-level function instead?

File: tests/chainer_tests/iterators_tests/test_multiprocess_iterator.py [hunk 2]
```diff
@@ -371,6 +372,34 @@ class TestMultiprocessIteratorSlow(
...
+    def setUp(self):
+        self.n_processes = 2
+        self.options = {'n_processes': self.n_processes,
+                        'n_prefetch': self.n_prefetch,
+                        'shared_mem': self.shared_mem}
+        if self.order_sampler is not None:
+            self.options.update(
+                {'shuffle': None, 'order_sampler': self.order_sampler})
+
+    def test_iterator_pickle(self):
```
Comment (niboshi) [inline tests/chainer_tests/iterators_tests/test_multiprocess_iterator.py]: How about taking some before pickling / taking the remaining after unpickling?

File: chainer/iterators/multiprocess_iterator.py [hunk 3]
```diff
@@ -129,7 +128,10 @@ def __init__(self, dataset, batch_size, repeat=True, shuffle=None,
                 order_sampler = ShuffleOrderSampler()
         self.order_sampler = order_sampler
 
-        self._comm = _Communicator(self.n_prefetch, dataset_timeout)
+        self.initialize_loop()
+
+    def initialize_loop(self):
```
Comment (niboshi) [inline chainer/iterators/multiprocess_iterator.py]: This should be private. Prefix with `_`.

File: chainer/iterators/multiprocess_iterator.py [hunk 4]
```diff
@@ -261,6 +263,40 @@ def _epoch_size(self):
...
+
+        # Unpickling resets the instance without calling __init__
+        # Chainer serializers dumps the state in an existing
+        # object hence we need to save the initial parameters too
+        init = self.__dict__.copy()
+        del init['_comm']
+        del init['_state']
+        del init['_prefetch_loop']
+
+        # TODO: when pickling in one process and restoring in other
```
Comment (niboshi) [inline chainer/iterators/multiprocess_iterator.py]: Add your name after `TODO`.

Comment (niboshi) [inline chainer/iterators/multiprocess_iterator.py]: Please check the grammar of the comment.

File: chainer/iterators/multiprocess_iterator.py [hunk 5]
```diff
@@ -425,6 +459,12 @@ def _allocate_shared_memory(self):
                 sharedctypes.RawArray('b', self.batch_size * self.mem_size)
 
     def launch_thread(self):
+        # Defer memory allocation due to pickling issues.
```
Comment (niboshi) [inline chainer/iterators/multiprocess_iterator.py]: It looks this comment is about "the change", not "the code".
This call itself does not defer the memory allocation.
How about something like "Memory allocation is deferred until the launch of the threads due to ..."?

Comment (emcastillo) [reply chainer/iterators/multiprocess_iterator.py]: With the new changes, deferring the memory allocation was not needed, so I reverted it :)

**Comment by emcastillo** [comment]
Addressed the comments. Thanks!

**Comment by niboshi** [comment]
Thank you for the fix.
Jenkins, test this please

**Pull request merged**